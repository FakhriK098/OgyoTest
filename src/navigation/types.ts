import { NavigatorScreenParams } from '@react-navigation/native';

export type RootStackParamList = {
  MainTabs: NavigatorScreenParams<MainTabParamList>;
  Detail: { id: string; title: string; data?: any };
};

export type HomeStackParamList = {
  Detail: { id: string };
};

export type MainTabParamList = {
  Home: undefined;
  User: undefined;
};
