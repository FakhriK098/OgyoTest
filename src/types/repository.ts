export interface IOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IRepository {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IOwner;
  html_url: string;
  description: string;
}

export interface IFetchRepositoriesPayload {
  since?: number;
}

export interface ISearchRepositoriesPayload {
  query: string;
}

export interface ILinkHeader {
  next?: string;
  prev?: string;
  first?: string;
  last?: string;
}
