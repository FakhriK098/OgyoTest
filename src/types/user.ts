export interface IUser {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IUserOwner;
  html_url: string;
  description: string;
  size: number;
  stargazers_count: number;
  watchers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: ILicense;
  topics: string[];
  network_count: number;
  subscribers_count: number;
}

export interface ILicense {
  key: string;
  name: string;
  spdx_id: string;
  url: string;
  node_id: string;
}

export interface IUserOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IFetchUserPayload {
  fullName?: string;
}

export interface ILabelProps {
  title: string;
  value?: string;
}
