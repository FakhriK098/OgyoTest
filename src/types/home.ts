import { IRepository } from './repository';

export interface IFloatingActionButtonProps {
  onPress: () => void;
}

export interface ISearchBarProps {
  onSearch: (query: string) => void;
  onSelectSuggestion: (repository: IRepository) => void;
  onClear: () => void;
  suggestions: IRepository[];
  loading: boolean;
  placeholder?: string;
  onSubmitSearch: (query: string) => void;
}

export type TSortOrder = 'ASC' | 'DESC';

export interface ISortFilterModalProps {
  visible: boolean;
  onClose: () => void;
  onSort: (order: TSortOrder) => void;
  currentOrder: TSortOrder;
}
