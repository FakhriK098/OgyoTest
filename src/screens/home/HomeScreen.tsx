import React, { useEffect, useState, useMemo } from 'react';
import {
  View,
  Text,
  StyleSheet,
  SafeAreaView,
  FlatList,
  ActivityIndicator,
  RefreshControl,
} from 'react-native';
import { colors } from '../../themes/colors';
import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import {
  fetchRepositoriesRequest,
  fetchMoreRepositoriesRequest,
  fetchSearchRepositoriesRequest,
  clearSearchResults,
} from '../../store/slices/repositorySagaSlice';
import { IRepository } from '../../types/repository';
import CardRepository from './components/CardRepository';
import ItemSeparator from '../../components/ItemSeparator';
import FloatingActionButton from './components/FloatingActionButton';
import SearchBar from './components/SearchBar';
import { TSortOrder } from '../../types/home';
import SortFilterModal from './components/SortFilterModal';
import { ASC } from '../../utils/constants';

const HomeScreen = () => {
  const dispatch = useAppDispatch();
  const {
    repositories,
    searchRepositories,
    loading,
    loadingMore,
    searchLoading,
    error,
    hasMore,
  } = useAppSelector(state => state.repository);

  const [isSearching, setIsSearching] = useState(false);
  const [showSearchResults, setShowSearchResults] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [sortOrder, setSortOrder] = useState<TSortOrder>(ASC);
  const [modalVisible, setModalVisible] = useState(false);

  useEffect(() => {
    dispatch(fetchRepositoriesRequest({ since: 0 }));
  }, [dispatch]);

  const sortedData = useMemo(() => {
    const dataToSort = showSearchResults ? searchRepositories : repositories;
    const sorted = [...dataToSort].sort((a, b) => {
      if (sortOrder === ASC) {
        return a.name.localeCompare(b.name);
      } else {
        return b.name.localeCompare(a.name);
      }
    });
    return sorted;
  }, [showSearchResults, searchRepositories, repositories, sortOrder]);

  const handleRefresh = () => {
    if (showSearchResults) {
      handleSubmitSearch(searchQuery);
    } else {
      dispatch(fetchRepositoriesRequest({ since: 0 }));
    }
  };

  const handleLoadMore = () => {
    if (!showSearchResults && !loadingMore && hasMore && !loading) {
      dispatch(fetchMoreRepositoriesRequest());
    }
  };

  const handleSearch = (query: string) => {
    setSearchQuery(query);
    if (query.trim().length > 0) {
      setIsSearching(true);
      dispatch(fetchSearchRepositoriesRequest({ query }));
    }
  };
  const handleSubmitSearch = (query: string) => {
    setSearchQuery(query);
    setShowSearchResults(true);
    setIsSearching(true);
    if (query.trim().length > 0) {
      dispatch(fetchSearchRepositoriesRequest({ query }));
    }
  };

  const handleSelectSuggestion = (repository: IRepository) => {
    setSearchQuery(repository.name);
    setShowSearchResults(true);
    setIsSearching(true);
    dispatch(fetchSearchRepositoriesRequest({ query: repository.name }));
  };

  const handleClearSearch = () => {
    setIsSearching(false);
    setShowSearchResults(false);
    setSearchQuery('');
    dispatch(clearSearchResults());
  };

  const handleOpenModal = () => {
    setModalVisible(true);
  };

  const handleCloseModal = () => {
    setModalVisible(false);
  };

  const handleSort = (order: TSortOrder) => {
    setSortOrder(order);
  };

  const renderItem = ({ item }: { item: IRepository }) => (
    <CardRepository item={item} />
  );

  const renderFooter = () => {
    if (!loadingMore) return null;
    return (
      <View style={styles.footerLoader}>
        <ActivityIndicator size="small" color={colors.shade800} />
      </View>
    );
  };

  const renderEmpty = () => {
    if (loading && !showSearchResults) {
      return (
        <View style={styles.emptyContainer}>
          <ActivityIndicator size="large" color={colors.shade800} />
          <Text style={styles.loadingText}>Loading repositories...</Text>
        </View>
      );
    }

    if (showSearchResults && searchLoading) {
      return (
        <View style={styles.emptyContainer}>
          <ActivityIndicator size="large" color={colors.shade800} />
          <Text style={styles.loadingText}>Searching...</Text>
        </View>
      );
    }

    if (
      showSearchResults &&
      searchRepositories.length === 0 &&
      !searchLoading
    ) {
      return (
        <View style={styles.emptyContainer}>
          <Text style={styles.noDataText}>No Data Found</Text>
        </View>
      );
    }

    if (error) {
      return (
        <View style={styles.emptyContainer}>
          <Text style={styles.errorText}>{error}</Text>
        </View>
      );
    }

    return null;
  };

  return (
    <SafeAreaView style={styles.container}>
      <SearchBar
        onSearch={handleSearch}
        onSelectSuggestion={handleSelectSuggestion}
        onClear={handleClearSearch}
        onSubmitSearch={handleSubmitSearch}
        suggestions={
          isSearching && !showSearchResults ? searchRepositories : []
        }
        loading={searchLoading}
        placeholder="Search GitHub repository..."
      />

      <FlatList
        data={sortedData}
        renderItem={renderItem}
        keyExtractor={item => item.id.toString()}
        ItemSeparatorComponent={ItemSeparator}
        showsVerticalScrollIndicator={false}
        style={styles.flatList}
        refreshControl={
          <RefreshControl
            refreshing={
              (loading && repositories.length === 0) ||
              (searchLoading && showSearchResults)
            }
            onRefresh={handleRefresh}
          />
        }
        onEndReached={handleLoadMore}
        onEndReachedThreshold={0.5}
        ListFooterComponent={renderFooter}
        ListEmptyComponent={renderEmpty}
      />

      <FloatingActionButton onPress={handleOpenModal} />

      <SortFilterModal
        visible={modalVisible}
        onClose={handleCloseModal}
        onSort={handleSort}
        currentOrder={sortOrder}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.brokenWhite,
  },
  header: {
    backgroundColor: colors.white,
    padding: 8,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  flatList: {
    paddingVertical: 8,
    paddingHorizontal: 8,
  },
  footerLoader: {
    paddingVertical: 20,
    alignItems: 'center',
  },
  emptyContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 50,
  },
  errorText: {
    fontSize: 16,
    color: colors.shade800,
    textAlign: 'center',
  },
  loadingText: {
    marginTop: 10,
    fontSize: 14,
    color: colors.shade75,
  },
  noDataText: {
    fontSize: 20,
    fontWeight: '600',
    color: colors.shade800,
    marginBottom: 8,
  },
  noDataSubtext: {
    fontSize: 14,
    color: colors.shade75,
    textAlign: 'center',
  },
});

export default HomeScreen;
