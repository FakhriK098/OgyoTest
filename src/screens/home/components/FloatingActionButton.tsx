import React from 'react';
import { TouchableOpacity, StyleSheet, Text, View } from 'react-native';
import { colors } from '../../../themes/colors';
import { IFloatingActionButtonProps } from '../../../types/home';

const FloatingActionButton: React.FC<IFloatingActionButtonProps> = ({
  onPress,
}) => {
  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={styles.fab}
        onPress={onPress}
        activeOpacity={0.8}
      >
        <Text style={styles.text}>Filter Sort</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  fab: {
    borderRadius: 28,
    width: 100,
    backgroundColor: colors.white,
    justifyContent: 'center',
    alignItems: 'center',
    elevation: 8,
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,
    padding: 8,
  },
  text: {
    fontSize: 16,
    color: colors.black,
  },
  container: {
    position: 'absolute',
    bottom: 20,
    right: 0,
    left: 0,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default FloatingActionButton;
