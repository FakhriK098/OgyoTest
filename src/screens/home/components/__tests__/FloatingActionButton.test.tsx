import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import FloatingActionButton from '../FloatingActionButton';
import { IFloatingActionButtonProps } from 'src/types/home';

describe('FloatingActionButton', () => {
  const defaultProps: IFloatingActionButtonProps = {
    onPress: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with default props', () => {
    const { getByText } = render(<FloatingActionButton {...defaultProps} />);

    expect(getByText('Filter Sort')).toBeTruthy();
  });

  it('should handle press event', () => {
    const onPressMock = jest.fn();
    const props = { onPress: onPressMock };

    const { getByText } = render(<FloatingActionButton {...props} />);

    const button = getByText('Filter Sort');
    fireEvent.press(button);

    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it('should be touchable and accessible', () => {
    const { getByText, UNSAFE_getByType } = render(
      <FloatingActionButton {...defaultProps} />,
    );

    const TouchableOpacity = require('react-native').TouchableOpacity;
    const touchable = UNSAFE_getByType(TouchableOpacity);
    expect(touchable).toBeTruthy();

    expect(getByText('Filter Sort')).toBeTruthy();
  });

  it('should match snapshot', () => {
    const { toJSON } = render(<FloatingActionButton {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with different onPress handler', () => {
    const customOnPress = jest.fn();
    const { toJSON } = render(<FloatingActionButton onPress={customOnPress} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
