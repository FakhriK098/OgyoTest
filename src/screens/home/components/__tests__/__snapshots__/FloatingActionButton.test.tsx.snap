// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingActionButton should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "bottom": 20,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 28,
        "elevation": 8,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 8,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 4.65,
        "width": 100,
      }
    }
  >
    <Text
      style={
        {
          "color": "#000000",
          "fontSize": 16,
        }
      }
    >
      Filter Sort
    </Text>
  </View>
</View>
`;

exports[`FloatingActionButton should match snapshot with different onPress handler 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "bottom": 20,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 28,
        "elevation": 8,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 8,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 4.65,
        "width": 100,
      }
    }
  >
    <Text
      style={
        {
          "color": "#000000",
          "fontSize": 16,
        }
      }
    >
      Filter Sort
    </Text>
  </View>
</View>
`;
